Smdb multiple authentication methods

JoeArmani opened the pull request:
This PR addresses multiple login options in SMDB.
It is a newly requested update to Abhinav's switch login mechanism [update #493](https://github.com/episphere/dashboard/issues/493) 

-Participant can have max 1 email and 1 phone auth method
-Only one method is required
-phone auth can only be removed if email auth exists
-email auth can only be removed if phone auth exists
-both firebase Auth and the participant’s firestore profiles are updated

Important workaround:
There is an issue where Firebase Auth doesn’t allow email removal.
The ops team wanted email removal to be an option for participants.
To work around this: on email removal request, we:
—update the participant `firebase auth email` to `noReply${uid}@episphere.github.io]`
—update the firestore participant’s `firestore profile` to the same
—filter emails starting with `noReply` and return an empty string with status=`Not in use` 
This effectively disables the participant’s prior email auth method and was approved/requested by the ops team.

jhflorey commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -82,20 +82,20 @@ const fieldValues = {
+    else return variableValue;
 }
 
 export const formatInputResponse = (participantValue) => {
```
It looks like we no longer use `formatInputResponse`

jhflorey commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -338,14 +331,6 @@ function isPhoneNumberPresent(participant, changedOption, conceptId, phoneType)
     return (participant[phoneType] || changedOption[phoneType]) && changedOption[phoneType] !== '' && conceptId != phoneType;
```
Should we use `conceptId !== phoneType` instead of `conceptId != phoneType`

JoeArmani replied:
I prefer the loose comparison here because `conceptId` comes in as a string and `phoneType` is a number. Here's a log statement with typeof attached when I edit `otherPhoneNumber` conceptId `793072415`. Otherwise I could wrap conceptId `parseInt(conceptId)` or `phoneType.toString()`
<img width="508" alt="Screenshot 2023-06-12 at 3 34 01 PM" src="https://github.com/episphere/dashboard/assets/93854858/f571e2ce-00ef-47c3-b63d-a499a50a497d">

jhflorey commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -460,6 +445,304 @@ export const resetChanges = (participant, originalHTML, bearerToken) => {
+            };
+            const inputFields = generateLoginFormInputFields(currentLogins, loginType);
+            const formButtons = generateAuthenticationFormButtons(!!(currentLogins.email && !currentLogins.email.startsWith('noReply')), !!currentLogins.phone, loginType);
+            const template = `
```
Should we use directly `body.innerHTML` instead of define `template`?

JoeArmani replied:
Definitely. Updated.

jhflorey commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -460,6 +445,304 @@ export const resetChanges = (participant, originalHTML, bearerToken) => {
+
+const postLoginData = async (url = '', data = {}, bearerToken) => {
+    try {
+        const response = await fetch(url, {
```
Could we return directly await?

JoeArmani replied:
Definitely. I was logging `response` in debugging and didn't consider this update. Done - thanks.

abhinavjonnada82 commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -460,6 +437,300 @@ export const resetChanges = (participant, originalHTML, bearerToken) => {
+ * for login removal, switch package and changedOption are generated from the participant object
+ */
+export const attachUpdateLoginMethodListeners = (participantAuthenticationEmail, participantAuthenticationPhone, participantToken, participantUid, bearerToken) => {
+    const createListener = (loginType) => {
```
Like this approach!

abhinavjonnada82 reviewed (approved):
looks good..

anthonypetersen reviewed (approved):
just a few comments, looks good

abhinavjonnada82 commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -82,38 +82,29 @@ const fieldValues = {
-    }
+    if (!variableValue) return '';
+    else if (conceptId === 'Change Login Email' && variableValue.startsWith('noreply')) return '';
+    else if (variableValue in fieldValues) return fieldValues[variableValue];
```
i had the same question but looks @JoeArmani defined it over [here](https://github.com/episphere/dashboard/blob/dev/siteManagerDashboard/participantDetailsHelpers.js#L70)

JoeArmani replied:
Nope, you're not missing it. I just pulled it out of `getFieldValues()` so it doesn't rebuild every time the `getFieldValues()` function is called because it's called for each row ~25x per form load. Now `fieldValues` sits directly above `getFieldValues` ~line 70.

```javascript
const fieldValues = {     
    [fieldMapping.yes]: 'Yes',
    [fieldMapping.no]: 'No',
    [fieldMapping.jr]: 'Jr.',
    [fieldMapping.sr]: 'Sr.',
    [fieldMapping.one]: 'I',
    [fieldMapping.two]: 'II',
    [fieldMapping.three]: 'III',
    [fieldMapping.second]: '2nd',
    [fieldMapping.third]: '3rd',
    [fieldMapping.prefPhone]: 'Phone',
    [fieldMapping.email]: 'Email',
}
```

JoeArmani commented on siteManagerDashboard/participantDetails.js:
```diff
@@ -56,8 +41,7 @@ export const renderParticipantDetails = (participant, changedOption, bearerToken
     renderReturnSearchResults();
-    updateUserSigninMechanism(participant, bearerToken);
-    updateUserLogin(participant, bearerToken);
+    attachUpdateLoginMethodListeners(participant[fieldMapping.accountEmail], participant[fieldMapping.accountPhone], participant.token, participant.state.uid, bearerToken);
```
There is no reason. Good catch - updated.
-removed from `index.js`
-removed from `participantDetails.js`
-`bearerToken` is now obtained in the try block of `processParticipantLoginMethod` (much narrower scope)

JoeArmani commented:
Update: @anthonypetersen @abhinavjonnada82 @jhflorey
I had to update `participantHeader.js` and `getImportantRows()` in `participantDetailsHelpers.js` due to conflicts we discovered today with the new `dataDestruction()` routines. The header was throwing `undefined` errors in a couple spots and the edit functionality needed to be removed for participants where dataDestruction has been executed. Here are the updated views for
(1) a participant after dataDestroy(), and
(2) an active participant.
Could I get a spot review on this before merging, please?
<img width="772" alt="SMDB - data destroyed participant" src="https://github.com/episphere/dashboard/assets/93854858/65879bf2-ac43-4c9d-b9a2-767b36811d05">
<img width="770" alt="SMDB - active, verified participant" src="https://github.com/episphere/dashboard/assets/93854858/071e4359-8834-47ea-a458-865c587b6742">

anthonypetersen commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -1193,6 +1193,10 @@ const processUserDataUpdate = async (changedUserDataForProfile, changedUserDataF
             changedUserDataForProfile[fieldMapping.userProfileHistory] = updateUserHistory(changedUserDataForHistory, userHistory, adminEmail);
         }
+        
+        if (changedUserDataForProfile[fieldMapping.fName]) changedUserDataForProfile['query.firstName'] = changedUserDataForProfile[fieldMapping.fName];
```
Does this render any changes to Connect Faas pointless?

JoeArmani replied:
I am not certain because I have not had a chance to inspect the search function in SMDB (Abhinav is on that). We brought up these changes yesterday, but updating query.firstName and query.lastName doesn't seem to be connecting with the SMDB search function at this moment.

JoeArmani replied:
Spoke on teams call, responding here for consistency: the specific changes  @anthonypetersen is referring to in connectFaas have been canceled.

anthonypetersen reviewed (approved):
looks good

JoeArmani commented:
Update:
-now handling query.allEmails. and query.allPhoneNo to support search
-added 'Save any User Profile edits before making Login Changes' row for better and safer UX

JoeArmani commented:
Updated per final round of Abhinav comments:

-removed emailTypes duplicate declaration.
-added to documentation for `handleAllPhoneNoField()` and `handleAllEmailField()`. Located above each function in block comments.
-added `.trim()?` in front of `.toLowerCase()` to `query.firstName`, `query.lastName`,  and all `email` fields.

JoeArmani commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -1116,6 +1131,61 @@ export const submitClickHandler = async (participant, changedOption) => {
+   */
+  const handleAllEmailField = (changedUserDataForProfile, participant) => {
+    const allEmails = participant.query.allEmails ?? [];
+    const emailTypes = [fieldMapping.prefEmail, fieldMapping.email1, fieldMapping.email2];
```
updated, marking resolved

JoeArmani commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -1194,8 +1264,8 @@ const processUserDataUpdate = async (changedUserDataForProfile, changedUserDataF
         
-        if (changedUserDataForProfile[fieldMapping.fName]) changedUserDataForProfile['query.firstName'] = changedUserDataForProfile[fieldMapping.fName];
-        if (changedUserDataForProfile[fieldMapping.lName]) changedUserDataForProfile['query.lastName'] = changedUserDataForProfile[fieldMapping.lName];
+        if (changedUserDataForProfile[fieldMapping.fName]) changedUserDataForProfile['query.firstName'] = changedUserDataForProfile[fieldMapping.fName].toLowerCase();
```
updated, marking resolved

JoeArmani commented on siteManagerDashboard/participantDetailsHelpers.js:
```diff
@@ -1116,6 +1131,61 @@ export const submitClickHandler = async (participant, changedOption) => {
+        allPhoneNo.push(changedUserDataForProfile[phoneType]);
+      }
+  
+      if (changedUserDataForProfile[phoneType] || changedUserDataForProfile[phoneType] === '') {
```
updated, marking resolved
